eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php

```diff
@@ -164,19 +164,44 @@ public function linkUrl($urlId, $fieldId, $versionNo)
      * @param int $versionNo
      * @param int[] $excludeUrlIds
      */
-    public function unlinkUrl($fieldId, $versionNo, array $excludeUrlIds = [])
+    public function unlinkUrl($fieldId, $versionNo, array $excludeUrlIds = []): void
     {
-        $deleteQuery = $this->connection->createQueryBuilder();
+        $selectQuery = $this->connection->createQueryBuilder();
+        $selectQuery
+            ->select('link.url_id')
+            ->from($this->connection->quoteIdentifier(self::URL_LINK_TABLE), 'link')
+            ->where(
+                $selectQuery->expr()->andX(
+                    $selectQuery->expr()->in(
+                        'link.contentobject_attribute_id',
+                        ':contentobject_attribute_id'
+                    ),
+                    $selectQuery->expr()->in(
+                        'link.contentobject_attribute_version',
+                        ':contentobject_attribute_version'
+                    )
+                )
+            )
+            ->setParameter(':contentobject_attribute_id', $fieldId, ParameterType::INTEGER)
+            ->setParameter(':contentobject_attribute_version', $versionNo, ParameterType::INTEGER);
 
+        $statement = $selectQuery->execute();
+        $potentiallyOrphanedUrls = $statement->fetchAll(PDO::FETCH_COLUMN);
+
+        if (empty($potentiallyOrphanedUrls)) {
+            return;
+        }
+
+        $deleteQuery = $this->connection->createQueryBuilder();
         $deleteQuery
             ->delete($this->connection->quoteIdentifier(self::URL_LINK_TABLE))
             ->where(
-                $deleteQuery->expr()->andX(
-                    $deleteQuery->expr()->in(
+                $selectQuery->expr()->andX(
```

Comment:
Why would you use `$selectQuery` expressions inside `$deleteQuery`? The expression builder is the same, but semantically it does not make sense to me.